Enum Support - Java Code Generation

TanayParikh opened the pull request:
- Added Enum code generation support for Java
- Updated src/foam/java/Enum.js starter file
- Added refinements in src/foam/java/refinements.js
- Updated  src/foam/java/Class.js and src/foam/java/Method.js to facilitate enum definitions

TanayParikh committed:
check in of enum support

TanayParikh committed:
Updated refinement to permit definitions in instances

TanayParikh committed:
Refactored value property to ordinal

- Added forLabel to get enum by label

TanayParikh committed:
Removed debugging code

TanayParikh committed:
Removed redundant EnumModel type definition

TanayParikh committed:
Merge branch 'master' into enum_support

kgrgreer commented:
@adamvy Can you please review. Thanks

adamvy commented:
@kgrgreer Only if you review #318 :)

adamvy commented on src/foam/java/Enum.js:
```diff
@@ -1,73 +1,55 @@
   }
 %>
 
   private final int index_;
```
Do we really need index_ and ordinal_ ?  Ordinal should be sufficient.

TanayParikh replied:
Thanks, that's a good point. I've removed the index property.

adamvy commented on src/foam/java/Enum.js:
```diff
@@ -1,73 +1,55 @@
-// Test with: foam.u2.ControllerMode.model_.javaSource();
+// Defines template for the EnumModel
```
Overall I wonder if we can do some better code sharing by making a foam.java.Enum class, and having it extend foam.java.Class

An Enum is in essence a Class with a fixed number of implementations.  If we extend foam.java.Class and update the outputJava method to output the VALUES, as well as whatever tweaks we need to match java enum syntax, we should also get methods/fields for free.


I would start by defining foam.java.Enum as a foam.CLASS() that extends foam.java.Class, and then trying to get the output correct.  Then refine buildJavaClass() onto foam.core.EnumModel, which should build a foam.java.Enum.

TanayParikh replied:
Hi Adam,

I've made foam.java.Enum and extended it from foam.java.Class. I've also updated the general code generation logic to move away from templates and more along the lines of the FOAM2 structure.

TanayParikh committed:
Removed index property

kgrgreer commented on src/foam/java/Enum.js:
```diff
@@ -1,97 +1,64 @@
-package <%= this.package %>;
 <%
 var labelForValue = function(value) {
   if (this.androidResource) {
```
Do we need this now? When we do, can we integrate it with FOAM's Message axiom support and then integrate the message support with Android Resources.

TanayParikh replied:
The androidResource is directly from the FOAM 1 enum template. Should I just go ahead and remove it?

kgrgreer replied:
Yes, just remove for now.

TanayParikh replied:
Removed

TanayParikh committed:
Removed androidResource check

TanayParikh committed:
Updated Java Enum support through extension of foam.CLASS

- Removed template for java enum
- Replaced with FOAM2 convention for java class generation

TanayParikh committed:
Removed debugging code

adamvy commented on src/foam/java/Enum.js:
```diff
@@ -1,106 +1,166 @@
 foam.CLASS({
-  refines: 'foam.core.EnumModel',
+  package: 'foam.java',
+  name: 'EnumModel',
```
This can just be called Enum, same way foam.java.Class is call Class not ClassModel.

adamvy commented on src/foam/java/Enum.js:
```diff
@@ -1,106 +1,166 @@
 
-  methods: [
+  properties: [
+    ['javaType', 'java.lang.Enum'],
```
Are these used anywhere?  I'm pretty sure these properties should be refined on foam.core.Enum

adamvy replied:
Ah reading further, you did refine them onto foam.core.Enum, which is good, these properties here should probably be deleted.

adamvy commented on src/foam/java/Enum.js:
```diff
@@ -1,106 +1,166 @@
+      o.out('}');
+    },
+
+    function toJavaSource() {
```
This should be inherited from foam.java.Class

adamvy commented on src/foam/java/refinements.js:
```diff
@@ -483,13 +484,66 @@ foam.CLASS({
+foam.CLASS({
+  refines: 'foam.core.AbstractEnum',
+
+  properties: [
```
I don't believe these properties are needed here.

adamvy commented on src/foam/java/refinements.js:
```diff
@@ -483,13 +484,66 @@ foam.CLASS({
+foam.CLASS({
+  refines: 'foam.core.EnumModel',
+
+  properties: [
```
Again with these properties :), I don't think you need them here either

adamvy commented on src/foam/java/Enum.js:
```diff
@@ -1,106 +1,166 @@
+    ['javaJSONParser', 'foam.lib.json.FObjectParser'],
   ],
+  methods: [
+    function outputJava(o) {
```
This is a good start.

Next I would look at if you can refactor this to share some code with outputJava in foam.java.Class.  There is a fair amount of duplication here that can probably be cleaned up.

TanayParikh committed:
Refactored Java `EnumModel` to `Enum`

TanayParikh committed:
Removed redundant properties

TanayParikh committed:
Removed toJavaSource function (inherited)

TanayParikh committed:
Updated java/Method.js to facilitate constructor

- Constructor doesn’t have a set type
- Is there an existing best convention for this?

TanayParikh committed:
Added ability to create enum java code to foam.java.class

TanayParikh committed:
Updated foam.java.enum to utilize foam.java.class

- Increase code re-use
- Added isEnum property to distinguish enums from classes

kirkeaton commented:
@adamvy Are there any more changes required for this PR?

adamvy commented:
Sorry for the delay, this looks good to me, thanks!
